### PR TITLE
lib/db: Dont recurse on flush (fixes #6905)

### LINKED ISSUE #6905: fatal error: stack overflow in lib/db
Sentry Issue: [SYNCTHING-37M](https://sentry.syncthing.net/syncthing/syncthing/issues/3343/?referrer=github_integration)

```
Stacktrace (aktuellster Aufruf als erstes):

  File "C:/BuildAgent/work/174e136266f8a219/lib/db/keyer.go", line 307, in GenerateFolderMetaKey
    func (k defaultKeyer) GenerateFolderMetaKey(key, folder []byte) (folderMetaKey, error) {
  File "<autogenerated>", line 1, in GenerateFolderMetaKey
  File "C:/BuildAgent/work/174e136266f8a219/lib/db/meta.go", line 84, in toDB
    key, err := m.keyer.GenerateFolderMetaKey(nil, folder)
  File "C:/BuildAgent/work/174e136266f8a219/lib/db/meta.go", line 77, in func1
    return m.toDB(t, folder)
  File "C:/BuildAgent/work/174e136266f8a219/lib/db/backend/leveldb_backend.go", line 188, in flush
    if err := hook(t); err != nil {
  File "C:/BuildAgent/work/174e136266f8a219/lib/db/backend/leveldb_backend.go", line 183, in checkFlush
    return t.flush()
  File "C:/BuildAgent/work/174e136266f8a219/lib/db/backend/leveldb_backend.go", line 159, in Put
    return t.checkFlush(dbFlushBatchMax)
  File "C:/BuildAgent/work/174e136266f8a219/lib/db/meta.go", line 100, in toDB
    err = t.Put(key, bs)
  File "C:/BuildAgent/work/174e136266f8a219/lib/db/meta.go", line 77, in func1
    return m.toDB(t, folder)
  File "C:/BuildAgent/work/174e136266f8a219/lib/db/backend/leveldb_backend.go", line 188, in flush
    if err := hook(t); err != nil {
...
(114 additional frame(s) were not displayed)

fatal error: stack overflow
```

### PR DESCRIPTION
**Pull request by AudriusButkevicius**

### CODE REVIEW

**Review by calmh — DISMISSED** [review]

**Comment by calmh** [comment]
I'm adding a test

**Comment by imsodin** [comment]
I think badger has the same issue, looking into it.

**Comment by calmh** [comment]
yes, and test shows it, will fix similarly

**Comment by calmh** [comment]
or actually no, the test just doesn't work on badger because we don't do auto flushes there...

**Review by calmh — DISMISSED** [review]

File: lib/db/backend/badger_backend.go [hunk 1]
```diff
@@ -255,6 +255,7 @@ type badgerTransaction struct {
 	rel         *releaser
 	size        int
 	commitHooks []CommitHook
+	inFlush     bool
```
Comment (imsodin) [inline lib/db/backend/badger_backend.go:258]: unused

Comment (calmh) [reply lib/db/backend/badger_backend.go:258]: ugh

**Review by imsodin — APPROVED** [review]

**Review by calmh — APPROVED** [review]

**Pull request merged**